pkg/controller/pcideviceclaim/pcideviceclaim_controller.go

```diff
@@ -277,43 +256,32 @@ func pciDeviceIsClaimed(pd *v1beta1.PCIDevice, pdcs *v1beta1.PCIDeviceClaimList)
 // A PCI Device is considered orphaned if it is bound to vfio-pci,
 // but has no PCIDeviceClaim. The assumption is that this controller
 // will manage all PCI passthrough, and consider orphaned devices invalid
-func (h Handler) unbindOrphanedPCIDevices(nodename string) error {
-	pdcs, err := h.pdcClient.List(metav1.ListOptions{})
-	if err != nil {
-		logrus.Errorf("Error listing PCI Device Claims: %s", err)
-		return err
-	}
-	pds, err := h.pdClient.List(metav1.ListOptions{})
-	if err != nil {
-		logrus.Errorf("Error listing PCI Devices: %s", err)
-		return err
-	}
-
+func getOrphanedPCIDevices(
+	nodename string,
+	pdcs *v1beta1.PCIDeviceClaimList,
+	pds *v1beta1.PCIDeviceList,
+) (*v1beta1.PCIDeviceList, error) {
+	pdsOrphaned := v1beta1.PCIDeviceList{}
 	for _, pd := range pds.Items {
 		isVfioPci := pd.Status.KernelDriverInUse == "vfio-pci"
 		isOnThisNode := nodename == pd.Status.NodeName
 		if isVfioPci && isOnThisNode && !pciDeviceIsClaimed(&pd, pdcs) {
-			logrus.Infof("PCI Device %s is bound to vfio-pci but has no Claim, attempting to unbind", pd.Status.Address)
-			err := unbindDeviceFromDriver(pd.Status.Address, vfioPCIDriver)
-			if err != nil {
-				logrus.Errorf("Error unbinding device from vfio-pci: %s", err)
-				return err
-			}
+			pdsOrphaned.Items = append(pdsOrphaned.Items, *pd.DeepCopy())
 		}
 	}
-	return nil
+	return &pdsOrphaned, nil
 }
 
 // After reboot, the PCIDeviceClaim will be there but the PCIDevice won't be bound to vfio-pci
-func (h Handler) rebindAfterReboot(nodename string) {
-	logrus.Infof("Rebinding after reboot on node: %s", nodename)
+func (h Handler) rebindAfterReboot() {
```

Comment:
Question: if `rebindAfterReboot` fails (e.g. `h.pdcClient.UpdateStatus` reports conflict), what will happen ? any way to retry ?

Reply:
Good idea, I can add a go routine that waits a minute and reruns the function: 

```go
		_, err = h.pdcClient.UpdateStatus(pdcCopy)
		if err != nil {
			logrus.Errorf("Failed to update PCIDeviceClaim status for %s: %s", pdc.Name, err)
			// Retry in 60 seconds
			go func() {
				time.Sleep(time.Second * 60)
				h.rebindAfterReboot()
			}()
		}
```

Reply:
well, if there are $n$ failures, then this will queue up $n$ copies, so I'll move this up to the calling function and return an error, and only retry if the `err` is not nil

Reply:
if I return err in the Register func, then `logrus.Fatal` will kill the process and Kubernetes will restart the Pod. That will implement the retry logic in a more Kubernetes-y way

Reply:
@w13915984028 https://github.com/harvester/pcidevices/pull/24